### PR TITLE
Fixes #21119: When a technique is not in the active techique tree, it can't be deleted in editor

### PR DESCRIPTION
**Pull request by ElaadF**
https://issues.rudder.io/issues/21119

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 1]
```diff
@@ -138,40 +134,56 @@ class TechniqueWriter (
       cr1.copy(directives = cr1.directives ++ cr2.directives)
     }
 
-    for {
-      directives <- readDirectives.getFullDirectiveLibrary().map(_.allActiveTechniques.values.filter(_.techniqueName.value == techniqueId.name.value).flatMap(_.directives).filter(_.techniqueVersion == techniqueId.version))
-      technique  <- techniqueRepository.get(techniqueId).notOptional(s"No Technique with ID '${techniqueId.toString()}' found in reference library.")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala:143]: You should extract the for loop in a function (maybe an inner function, it does not need to be global for now)

Same for the none part, it should be in an inner function

**Review by fanf — DISMISSED** [review]

**Comment by ElaadF** [comment]
I'm surprised I did refactor in functions, I had pushed the changes but didn't appear.. I will check locally

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala [hunk 2]
```diff
@@ -138,40 +139,66 @@ class TechniqueWriter (
...
+        _          <- archiver.deleteTechnique(technique.name, techniqueVersion, categories.map(_.id.name.value), modId,committer, s"Deleting technique ${technique.name}/${techniqueVersion}")
+        _          <- techLibUpdate.update(modId, committer, Some(s"Update Technique library after deletion of Technique ${technique.name}")).toIO.chainError(
+                        s"An error occurred during technique update after deletion of Technique ${technique.name}"
+                      )
+      } yield {
+        ()
+      }
+    }
+
+    def removeInvalidTechnique(basePath: String, techniqueName: String): IOResult[Unit] = {
```
Comment (ElaadF) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala:176]: @VinceMacBuche the inner function

**Review by VinceMacBuche — APPROVED** [review]

**Comment by ElaadF** [comment]
OK, squash merging this PR

**Pull request merged**